lib/cylc/remote.py

```diff
@@ -214,72 +304,589 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
+        command = construct_ssh_cmd(
+            cmd, user=self.owner, host=self.host, forward_x11=forward_x11,
+            ssh_login_shell=self.ssh_login_shell, ssh_cylc=self.ssh_cylc,
+            set_UTC=True, allow_flag_opts=True)
 
         if dry_run:
             return command
+        else:
+            return run_ssh_cmd(command)
+
+
+class EmptyHostList(Exception):
+    """Exception to be raised if there are no valid run hosts.
+
+       Raise if, from the global configuration settings, no hosts are listed
+       for potential appointment to run suites (as 'run hosts') or none
+       satisfy requirements to pass 'run host select' specifications.
+
+       Print message with the current state of relevant settings for info."""
+
+    def __str__(self):
+        msg = "No hosts currently compatible with this global configuration:\n"
+        suite_server_cfg_items = (['run hosts'], ['run host select', 'rank'],
+                                  ['run host select', 'thresholds'])
+        for cfg_end_ref in suite_server_cfg_items:
+            cfg_full_ref = cfg_end_ref.insert(0, 'suite servers')
+            # Add 2-space indentation for clarity in distinction of items.
+            msg = '\n  '.join([msg, cfg_item[-1] + ':',
+                               '  ' + glbl_cfg().get(cfg_full_ref)])
+        return msg
+
+
+class HostAppointer(object):
+    """Appoint the most suitable host to run a suite on.
+
+       Determine the one host most suitable to (re-)run a suite on from all
+       'run hosts' given the 'run host selection' ranking & threshold options
+       as specified (else taken as default) in the global configuration.
+    """
+
+    CMD_BASE = "get-host-metrics"  # 'cylc' prepended by remote_cylc_cmd.
+
+    def __init__(self):
+        self.use_disk_path = "/"
+        self.max_processes = 5
 
         try:
-            popen = Popen(command)
-        except OSError as exc:
-            sys.exit(r'ERROR: remote command invocation failed %s' % exc)
+            self.HOSTS = glbl_cfg().get(['suite servers', 'run hosts'])
+            self.RANK_METHOD = glbl_cfg().get(
+                ['suite servers', 'run host select', 'rank'])
+            self.THRESHOLDS = glbl_cfg().get(
+                ['suite servers', 'run host select', 'thresholds'])
+        except KeyError:  # Catch for unittest which updates these internally.
+            self.HOSTS = []
+            self.RANK_METHOD = 'random'
+            self.THRESHOLDS = None
+
+        self.PARSED_THRESHOLDS = self.parse_thresholds(self.THRESHOLDS)
+
+    @staticmethod
+    def parse_thresholds(raw_thresholds_spec):
+        """Parse raw 'thresholds' global configuration option to dict."""
+        if not raw_thresholds_spec:
+            return {}
+        valid_thresholds = {}
+        for threshold in raw_thresholds_spec.split(';'):
+            threshold = threshold.strip()
+            try:
+                measure, cutoff = threshold.split(' ')
```

Comment:
I would be tempted to do:

```python
measure, cutoff = threshold.split()
```

To provide resistance against someone putting multiple spaces in a threshold string.

Reply:
Good thinking.